### ISSUE TITLE
Implement proper deallocation

### ISSUE THREAD

**Issue by certik** [lcompilers/lpython#996]
Of:

* dicts
* lists
* strings (#805)
* function return values (#494)

And anything else that might leak memory currently.

**Comment by czgdp1807** [comment]
The deallocation should be implicit (done by LPython), explicit (done by user via `del` statements), right?

**Comment by certik** [comment]
Yes, implicit when the (allocatable) variable goes out of scope. I think dicts, lists, strings should have "allocatable" set, and then we should add some mechanism into libasr that both LFortran and LPython can use to automatically insert the "ImplicitDeallocate" ASR nodes.

And `del` would delete it right away.
